tests/tlm/test_concurrency.py

```diff
@@ -0,0 +1,140 @@
+    ):
+        async with tlm_rate_handler:
+            pass
+
+        assert (
+            tlm_rate_handler._congestion_window == expected_limit_value
+        ), "Congestion window is not increased linearly in congestion control"
+        assert (
+            tlm_rate_handler._send_semaphore._value == tlm_rate_handler._congestion_window
+        ), "Send semaphore value does not match congestion window in congestion control"
+
+
+@pytest.mark.parametrize("initial_congestion_window", [4, 5, 10, 101])
+@pytest.mark.asyncio
+async def test_rate_handler_rate_limit_error(
+    tlm_rate_handler: TlmRateHandler,
+    initial_congestion_window: int,
+) -> None:
+    """Tests rate handler decrease behavior on a rate limit error.
+
+    Expected behavior:
+    - Limiter decreases congestion window multiplicatively
+    - Limiter send semaphore value matches congestion window
+    - RateLimitError is raised (not suppressed by context manager)
+    """
+    # arrange -- set current congestion window
+    tlm_rate_handler._congestion_window = initial_congestion_window
+    tlm_rate_handler._send_semaphore._value = initial_congestion_window
+
+    # acquire rate limit and raise rate limit error, check that:
+    # - congestion window is decreased multiplicatively
+    # - send semaphore value matches congestion window
+    # - rate limit error is raised
+    with pytest.raises(RateLimitError):
+        async with tlm_rate_handler:
+            raise RateLimitError("", 0)
+
+        assert (
+            tlm_rate_handler._congestion_window
+            == initial_congestion_window // tlm_rate_handler.MULTIPLICATIVE_DECREASE_FACTOR
+        ), "Congestion window is not decreased multiplicatively in congestion avoidance"
+        assert (
+            tlm_rate_handler._send_semaphore._value == tlm_rate_handler._congestion_window
+        ), "Send semaphore value does not match congestion window in congestion avoidance"
+
+
+@pytest.mark.parametrize("initial_congestion_window", [4, 5, 10, 101])
+@pytest.mark.asyncio
+async def test_rate_handler_non_rate_limit_error(
+    tlm_rate_handler: TlmRateHandler,
+    initial_congestion_window: int,
+) -> None:
+    """Tests rate handler decrease behavior on a NON rate limit error.
+
+    Expected behavior:
+    - Limiter congestion window stays the same
+    - Limiter send semaphore value matches congestion window
+    - error is raised (not suppressed by context manager)
+    """
+    # arrange -- set current congestion window
+    tlm_rate_handler._congestion_window = initial_congestion_window
+    tlm_rate_handler._send_semaphore._value = initial_congestion_window
+
+    # acquire rate limit and raise rate limit error, check that:
+    # - congestion window is decreased multiplicatively
+    # - send semaphore value matches congestion window
+    # - rate limit error is raised
+    with pytest.raises(ValueError):
+        async with tlm_rate_handler:
+            raise ValueError
+
+        assert (
```

Comment:
these assertions need to be moved out of the with statement. when ValueError is raised it'll exit the block and not hit the assert statements. you can see this from me running the following tests below with an incorrect value
![Screenshot 2024-03-14 at 1 43 32 PM](https://github.com/cleanlab/cleanlab-studio/assets/13542329/9fe91698-65b2-4d3e-975b-4df56dd1ffc8)
![Screenshot 2024-03-14 at 1 44 00 PM](https://github.com/cleanlab/cleanlab-studio/assets/13542329/db778ad8-9af6-4f6c-a987-583fd3ced7b6)

Reply:
ah, and this is why we didn't get a failure for this [bug](https://github.com/cleanlab/cleanlab-studio/pull/201#discussion_r1524923733). excellent, full circle moment lol

this is fixed now too

Reply:
(these tests pass now w/ https://github.com/cleanlab/cleanlab-studio/pull/201/commits/8c1116d7337afb8e78a4cffd4b5c56d9ddcd7164)

Reply:
nice! much clearer now haha